lisa/tools/lspci.py

```diff
@@ -142,47 +158,107 @@ def _install(self) -> bool:
         return self._check_exists()
 
     def get_device_names_by_type(
-        self, device_type: str, force_run: bool = False
+        self, device_type: str, force_run: bool = False, use_pci_ids: bool = False
     ) -> List[str]:
         if device_type.upper() not in DEVICE_TYPE_DICT.keys():
             raise LisaException(f"pci_type '{device_type}' is not recognized.")
         class_names = DEVICE_TYPE_DICT[device_type.upper()]
         devices_list = self.get_devices(force_run)
-        devices_slots = [x.slot for x in devices_list if x.device_class in class_names]
+        devices_slots = []
+        if use_pci_ids:
+            for device in devices_list:
+                if (
+                    device.controller_id in CONTROLLER_ID_DICT[device_type.upper()]
+                    and device.vendor_id in VENDOR_ID_DICT[device_type.upper()]
+                    and device.device_id in DEVICE_ID_DICT[device_type.upper()]
+                ):
+                    devices_slots.append(device.slot)
+        else:
+            devices_slots = [
+                x.slot for x in devices_list if x.device_class in class_names
+            ]
         return devices_slots
 
     def get_devices_by_type(
-        self, device_type: str, force_run: bool = False
+        self, device_type: str, force_run: bool = False, use_pci_ids: bool = False
     ) -> List[PciDevice]:
         if device_type.upper() not in DEVICE_TYPE_DICT.keys():
             raise LisaException(
                 f"pci_type '{device_type}' is not supported to be searched."
             )
-        class_names = DEVICE_TYPE_DICT[device_type.upper()]
         devices_list = self.get_devices(force_run)
-        device_type_list = [x for x in devices_list if x.device_class in class_names]
+        device_type_list = []
+        if use_pci_ids:
+            for device in devices_list:
+                if (
+                    device.controller_id in CONTROLLER_ID_DICT[device_type.upper()]
+                    and device.vendor_id in VENDOR_ID_DICT[device_type.upper()]
+                    and device.device_id in DEVICE_ID_DICT[device_type.upper()]
+                ):
+                    device_type_list.append(device)
+        else:
+            class_names = DEVICE_TYPE_DICT[device_type.upper()]
+            device_type_list = [
+                x for x in devices_list if x.device_class in class_names
+            ]
         return device_type_list
 
+    @retry(KeyError, tries=10, delay=20)
```

Comment:
Why retry can fix key errors?

Reply:
Some times difference seen between outputs of "lspci -n" and "lspci -m" used in this routine.
It usually occurs while checking for VFs right after VM booting when the VF count is 8. 
All VF becomes available only after couple of minutes of VM booting.
This is for re-trying in such cases instead of simply failing.

Reply:
Can you check which one of "-n" or "-m" is incorrect, and why it's fixed by rerun? 10*20 means 200 seconds, if this method called many times, it slows down the test cases significantly.

Reply:
Both -n and -m are correct. Its not an issue of lspci commands. Its the number of PCI devices newly added between these command executions is causing the exception.